Add doc for diagnosing backpressure from Elasticsearch (#4097)
elastic/observability-docs: docs/en/observability/apm/apm-performance-diagnostic.asciidoc

[[apm-performance-diagnostic]]
=== APM Server performance diagnostic

[[apm-es-backpressure]]
[float]
==== Diagnosing backpressure from {es}

When {es} is under excessive load or indexing pressure, APM Server could experience the downstream backpressure when indexing new documents into {es}.

> Most commonly backpressure from {es} will manifest itself in a form of higher indexing latency and/or rejected requests, which in return could lead to denied incoming requests by APM Server.

Author: Added a note about throttling and timeouts on APM agents.

> Most commonly backpressure from {es} will manifest itself in a form of higher indexing latency and/or rejected requests, which in return could lead to denied incoming requests by APM Server.

Author: yes, looks good. Thank you @colleenmcginnis.

To quickly identify possible issues try looking for similar error logs lines in APM Server logs:

[source,json]
----
...
{"log.level":"error","@timestamp":"2024-07-27T23:46:28.529Z","log.origin":{"function":"github.com/elastic/go-docappender/v2.(*Appender).flush","file.name":"v2@v2.2.0/appender.go","file.line":370},"message":"bulk indexing request failed","service.name":"apm-server","error":{"message":"flush failed (429): [429 Too Many Requests]"},"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2024-07-27T23:55:38.612Z","log.origin":{"function":"github.com/elastic/go-docappender/v2.(*Appender).flush","file.name":"v2@v2.2.0/appender.go","file.line":370},"message":"bulk indexing request failed","service.name":"apm-server","error":{"message":"flush failed (503): [503 Service Unavailable] "},"ecs.version":"1.6.0"}
...
----

> To gain better insight into APM Server health and performance please consider enabling Elastic Stack monitoring by following the guide <<monitor>>.

Author: I used this wording because the root monitoring page https://www.elastic.co/guide/en/observability/current/apm-monitor-apm.html calls the monitoring feature this way, which made me slightly confused.

Updated per your suggestion now.

When enabled APM Server will additionally report a set of vital metrics to help you identify any performance degradation.

> Most notably metric fields `beats_stats.metrics.libbeat.output.events.*` will depict the number of documents that are pending in the buffer for indexing, already indexed successfully or failed during the indexing.

Author: I could do either option, let me know what is more preferable:
- only adding the doc for 8.15+
- adding for 8.15+ and split the second version for older stack versions

> Most notably metric fields `beats_stats.metrics.libbeat.output.events.*` will depict the number of documents that are pending in the buffer for indexing, already indexed successfully or failed during the indexing.

Reviewer: I prefer only adding for 8.15+ to avoid the extra PR, but interested to hear what others have to say.

See https://www.elastic.co/guide/en/beats/metricbeat/current/exported-fields-beat.html[{metricbeat} documentation] for the full list of exported metric fields.

> If rejected requests are identified try to follow the guide from https://www.elastic.co/guide/en/elasticsearch/reference/current/rejected-requests.html.

Author: Added the direct note about likely cause of undersized {es}.

Most commonly backpressure from {es} will manifest itself in a form of higher indexing latency and/or rejected requests, which in return could lead to denied incoming requests by APM Server.
As a result APM agents connected to the affected APM Server will suffer from throttling and/or request timeout when shipping APM events. 

To quickly identify possible issues try looking for similar error logs lines in APM Server logs:

[source,json]
----
...
{"log.level":"error","@timestamp":"2024-07-27T23:46:28.529Z","log.origin":{"function":"github.com/elastic/go-docappender/v2.(*Appender).flush","file.name":"v2@v2.2.0/appender.go","file.line":370},"message":"bulk indexing request failed","service.name":"apm-server","error":{"message":"flush failed (429): [429 Too Many Requests]"},"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2024-07-27T23:55:38.612Z","log.origin":{"function":"github.com/elastic/go-docappender/v2.(*Appender).flush","file.name":"v2@v2.2.0/appender.go","file.line":370},"message":"bulk indexing request failed","service.name":"apm-server","error":{"message":"flush failed (503): [503 Service Unavailable]"},"ecs.version":"1.6.0"}
...
----

> To gain better insight into APM Server health and performance please consider enabling the monitoring feature by following the guide from <<apm-monitor-apm-self-install>>.

Reviewer 2: ```suggestion
To gain better insight into APM Server health and performance, consider enabling the monitoring feature by following the steps in <<apm-monitor-apm-self-install,Monitor a Fleet-managed APM Server>>.
```

Author: updated

When enabled APM Server will additionally report a set of vital metrics to help you identify any performance degradation.

> Pay careful attention to the next metric fields:
> - `beats_stats.metrics.libbeat.output.events.active` that represents the number of buffered pending documents waiting for indexing;
> - `beats_stats.metrics.libbeat.output.events.acked` that represents the number of indexing operations that have completed successfully;
> - `beats_stats.metrics.libbeat.output.events.failed` that represents the number of indexing operations that failed, it includes all failures;
> - `beats_stats.metrics.libbeat.output.events.toomany` that represents the number of indexing operations that failed due to {es} responding with 429 Too many Requests;
> - `beat.stats.output.elasticsearch.bulk_requests.active` that represents the number of bulk indexers available for making bulk index requests;
> - `beat.stats.output.elasticsearch.bulk_requests.completed` that represents the number of already completed bulk requests;
> - `beats_stats.metrics.output.elasticsearch.indexers.active` that represents the number of active bulk indexers that are concurrently processing batches;
> See https://www.elastic.co/guide/en/beats/metricbeat/current/exported-fields-beat.html[{metricbeat} documentation] for the full list of exported metric fields.

Reviewer 2: This is [not currently rendering as a list](https://observability-docs_bk_4097.docs-preview.app.elstc.co/guide/en/observability/master/apm-performance-diagnostic.html).

<img width="880" alt="Screenshot 2024-07-31 at 1 02 54 PM" src="https://github.com/user-attachments/assets/78e363c6-c3ca-47ad-8a8f-c3f9002d624d">

Try this instead:

```suggestion
Pay careful attention to the next metric fields:

* `beats_stats.metrics.libbeat.output.events.active` that represents the number of buffered pending documents waiting for indexing;
* `beats_stats.metrics.libbeat.output.events.acked` that represents the number of indexing operations that have completed successfully;
* `beats_stats.metrics.libbeat.output.events.failed` that represents the number of indexing operations that failed, it includes all failures;
* `beats_stats.metrics.libbeat.output.events.toomany` that represents the number of indexing operations that failed due to {es} responding with 429 Too many Requests;
* `beat.stats.output.elasticsearch.bulk_requests.active` that represents the number of bulk indexers available for making bulk index requests;
* `beat.stats.output.elasticsearch.bulk_requests.completed` that represents the number of already completed bulk requests;
* `beats_stats.metrics.output.elasticsearch.indexers.active` that represents the number of active bulk indexers that are concurrently processing batches;

See https://www.elastic.co/guide/en/beats/metricbeat/current/exported-fields-beat.html[{metricbeat} documentation] for the full list of exported metric fields.
```

Reviewer: typo: `beats_stats.output.elasticsearch.bulk_requests.available`, not `beat.stats.output.elasticsearch.bulk_requests.active`. Should explicitly call out that if it is 0, it is facing ES backpressure.

Reviewer: Also, some metrics are using `beats_stats`, some are using `beat.stats`. Let's stick with beats_stats

Author: Thank you for catching this up @colleenmcginnis 🙏 I'm pretty new to asciidoc markup language, I'll try to run local asciidoc text processor to better preview my changes before committing them in the future.

Author: updated in the last commit

> One likely cause of excessive indexing pressure or rejected requests is undersized {es}, to mitigate try following the guide from https://www.elastic.co/guide/en/elasticsearch/reference/current/rejected-requests.html.

Reviewer 2: ```suggestion
One likely cause of excessive indexing pressure or rejected requests is undersized {es}. To mitigate this, follow the guidance in {ref}/rejected-requests.html[Rejected requests].
```

Author: updated

> Alternatively try adjusting `max_retries` and `timeout` settings from <<apm-elasticsearch-output>> to reduce APM Server indexing pressure, note however that reducing retries may lead to a higher rate of dropped APM events.

Reviewer 2: ```suggestion
Alternatively, you can try adjusting `max_retries` and `timeout` settings described in <<apm-elasticsearch-output,Configure the Elasticsearch output>> to reduce APM Server indexing pressure. However, reducing retries may lead to a higher rate of dropped APM events.
```

Author: updated

> Alternatively try adjusting `max_retries` and `timeout` settings from <<apm-elasticsearch-output>> to reduce APM Server indexing pressure, note however that reducing retries may lead to a higher rate of dropped APM events.

Reviewer: As we are mentioning max_retries and timeout, we may also want to mention how `flush_bytes` can be adjusted to potentially buffer more documents and produce a larger bulk request, at the expense of higher memory usage. This config is documented.

Also, "Alternatively" is a bit soft imo, and may imply that the workaround to increase flush_bytes or reduce max_retries is as good as scaling up ES. I believe it is better to state/imply that it is really just a workaround if scaling up ES is not an option.

Author: Updated according to the suggestions.

[[apm-performance-diagnostic]]
=== APM Server performance diagnostic

[[apm-es-backpressure]]
[float]
==== Diagnosing backpressure from {es}

When {es} is under excessive load or indexing pressure, APM Server could experience the downstream backpressure when indexing new documents into {es}.
Most commonly, backpressure from {es} will manifest itself in the form of higher indexing latency and/or rejected requests, which in return could lead APM Server to deny incoming requests.

> As a result APM agents connected to the affected APM Server will suffer from throttling and/or request timeout when shipping APM events. 

Reviewer: ```suggestion
As a result, APM agents connected to the affected APM Server will suffer from throttling and/or request timeout when shipping APM events. 
```

Author: updated

To quickly identify possible issues try looking for similar error logs lines in APM Server logs:

[source,json]
----
...
{"log.level":"error","@timestamp":"2024-07-27T23:46:28.529Z","log.origin":{"function":"github.com/elastic/go-docappender/v2.(*Appender).flush","file.name":"v2@v2.2.0/appender.go","file.line":370},"message":"bulk indexing request failed","service.name":"apm-server","error":{"message":"flush failed (429): [429 Too Many Requests]"},"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2024-07-27T23:55:38.612Z","log.origin":{"function":"github.com/elastic/go-docappender/v2.(*Appender).flush","file.name":"v2@v2.2.0/appender.go","file.line":370},"message":"bulk indexing request failed","service.name":"apm-server","error":{"message":"flush failed (503): [503 Service Unavailable]"},"ecs.version":"1.6.0"}
...
----

> To gain better insight into APM Server health and performance, consider enabling the monitoring feature by following the steps in <<apm-monitor-apm-self-install,Monitor a Fleet-managed APM Server>>.

Reviewer: ```suggestion
To gain better insight into APM Server health and performance, consider enabling the monitoring feature by following the steps in <<apm-monitor-apm,Monitor APM Server>>.
```

Author: updated

> When enabled APM Server will additionally report a set of vital metrics to help you identify any performance degradation.

Reviewer: ```suggestion
When enabled, APM Server will additionally report a set of vital metrics to help you identify any performance degradation.
```

Author: updated

Pay careful attention to the next metric fields:

* `beats_stats.metrics.libbeat.output.events.active` that represents the number of buffered pending documents waiting for indexing;

> (_if this value is increasing rapidly it indicates {es} backpressure_)

Reviewer: this I'm not 100% sure. Yes, if active is _increasing_, there _may_ be backpressure, but it _may_ also be due to increased ingestion.
```suggestion
(_if this value is increasing rapidly it may indicate {es} backpressure_)
```

* `beats_stats.metrics.libbeat.output.events.acked` that represents the number of indexing operations that have completed successfully;

> * `beats_stats.metrics.libbeat.output.events.failed` that represents the number of indexing operations that failed, it includes all failures;

Reviewer 3: nit

```suggestion
* `beats_stats.metrics.libbeat.output.events.failed` that represents the number of indexing operations that failed, including all failures;
```

Reviewer 3: @carsonip did I understand correctly that `beats_stats.metrics.libbeat.output.events.failed` is counting the ES _documents_ that failed, rather than _bulk requests that failed?
The suggestion is based on this assumption.

Reviewer 3: Documentation is not super clear 🤔

Reviewer 3: It was also source of an SDH, this one https://github.com/elastic/sdh-apm/issues/1367#issuecomment-2159748910

Reviewer: > @carsonip did I understand correctly that beats_stats.metrics.libbeat.output.events.failed is counting the ES documents that failed, rather than _bulk requests that failed?

correct
```
Failed:                 atomic.LoadInt64(&a.docsFailed),
```

Author: I used description from the metrics from the code directly, fixed the wording in the doc now, it should be better.

> (_if this value is increasing rapidly it indicates {es} backpressure_)

Reviewer: ```suggestion
(_if this value is increasing rapidly it may indicate {es} backpressure_)
```

Author: updated

* `beats_stats.metrics.libbeat.output.events.toomany` that represents the number of indexing operations that failed due to {es} responding with 429 Too many Requests;
(_if this value is increasing rapidly it indicates {es} backpressure_)
* `beats_stats.output.elasticsearch.bulk_requests.available` that represents the number of bulk indexers available for making bulk index requests;
(_if this value is equal to 0 it indicates {es} backpressure_)
* `beats_stats.output.elasticsearch.bulk_requests.completed` that represents the number of already completed bulk requests;
* `beats_stats.metrics.output.elasticsearch.indexers.active` that represents the number of active bulk indexers that are concurrently processing batches;

> See https://www.elastic.co/guide/en/beats/metricbeat/current/exported-fields-beat.html[{metricbeat} documentation] for the full list of exported metric fields.

Reviewer: @colleenmcginnis is there a way to link to the metricbeat docs on the same release version?

Reviewer 2: Yep! 

```suggestion
See {metricbeat-ref}/exported-fields-beat.html[{metricbeat} documentation] for the full list of exported metric fields.
```

Author: updated

One likely cause of excessive indexing pressure or rejected requests is undersized {es}. To mitigate this, follow the guidance in {ref}/rejected-requests.html[Rejected requests].

> If scaling {es} resources up is not an option, you can try to workaround by adjusting `flush_bytes`, `flush_interval`, `max_retries` and `timeout` settings described in <<apm-elasticsearch-output,Configure the Elasticsearch output>> to reduce APM Server indexing pressure.

Reviewer 3: ```suggestion
If scaling {es} resources up is not an option, you can adjust the `flush_bytes`, `flush_interval`, `max_retries` and `timeout` settings described in <<apm-elasticsearch-output,Configure the Elasticsearch output>> to reduce APM Server indexing pressure.
```

Should we also include suggestions on how to tune the values?
Or give a sort of reference config? Not sure if we can provide one that is valid for all customers though

Author: Added an example yaml config. As you said it's not applicable to specific situation but probably it's still nice to have a config reference example in the doc in front of you.

Reviewer: Tbh I am not a big fan of providing an example here. If I were the user I would be tempted to skip the advice about scaling up ES and conveniently apply this example config and continue to get a bad experience from our APM offering. But now that it is added to the PR, I don't feel strongly to remove it, as we clearly stated that "if scaling ES is not an option".

Author: I can see your point Carson, and how that might be tempting to just copy paste the config example instead of more direct and effective scale up of ES.  I feel like it's a matter of preference whether to introduce this section or not, since it has value of quickly providing a visual config reference but at the same time it might encourage users to skip reading other parts of this doc.

We probably should be consistent with the other documentation pages, do we usually add config example in other places in the documentation?

Also an alternative approach is to keep the example, but move it under a subsection "Workaround by configuring APM" at the bottom of the page where in the first sentence we state that it's not recommended to apply it, and only try to modify configs if the other options are exhausted.

Reviewer: sounds good, left a suggestion.

(_if this value is increasing rapidly it may indicate {es} backpressure_)
* `beats_stats.metrics.libbeat.output.events.toomany` that represents the number of documents that failed to ingest due to {es} responding with 429 Too many Requests;
(_if this value is increasing rapidly it may indicate {es} backpressure_)
* `beats_stats.output.elasticsearch.bulk_requests.available` that represents the number of bulk indexers available for making bulk index requests;
(_if this value is equal to 0 it may indicate {es} backpressure_)
* `beats_stats.output.elasticsearch.bulk_requests.completed` that represents the number of already completed bulk requests;
* `beats_stats.metrics.output.elasticsearch.indexers.active` that represents the number of active bulk indexers that are concurrently processing batches;

See {metricbeat-ref}/exported-fields-beat.html[{metricbeat} documentation] for the full list of exported metric fields.

One likely cause of excessive indexing pressure or rejected requests is undersized {es}. To mitigate this, follow the guidance in {ref}/rejected-requests.html[Rejected requests].

> If scaling {es} resources up is not an option, you can adjust the `flush_bytes`, `flush_interval`, `max_retries` and `timeout` settings described in <<apm-elasticsearch-output,Configure the Elasticsearch output>> to reduce APM Server indexing pressure. However, consider that increasing number of buffered documents and/or reducing retries may lead to a higher rate of dropped APM events. Down below a custom configuration example is listed where the number of default buffered documents is roughly doubled while {es} indexing retries are decreased simultaneously. This configuration provides a generic example and might not be applicable to your situation. Try adjusting the settings further to see what works for you.

Reviewer: As discussed, I agree that calling out as "Not recommended" and placing it in a separate paragraph should be clear enough
```suggestion
(Not recommended) If scaling {es} resources up is not an option, you can adjust the `flush_bytes`, `flush_interval`, `max_retries` and `timeout` settings described in <<apm-elasticsearch-output,Configure the Elasticsearch output>> to reduce APM Server indexing pressure. However, consider that increasing number of buffered documents and/or reducing retries may lead to a higher rate of dropped APM events. Down below a custom configuration example is listed where the number of default buffered documents is roughly doubled while {es} indexing retries are decreased simultaneously. This configuration provides a generic example and might not be applicable to your situation. Try adjusting the settings further to see what works for you.
```

Author: updated

> If scaling {es} resources up is not an option, you can adjust the `flush_bytes`, `flush_interval`, `max_retries` and `timeout` settings described in <<apm-elasticsearch-output,Configure the Elasticsearch output>> to reduce APM Server indexing pressure. However, consider that increasing number of buffered documents and/or reducing retries may lead to a higher rate of dropped APM events. Down bellow a custom configuration example is listed where the number of default buffered documents is roughly doubled while {es} indexing retries are decreased simultaneously. This configuration provides a generic example and might not be applicable to your situation. Try adjusting the settings further to see what works for you.

Reviewer 3: ```suggestion
If scaling {es} resources up is not an option, you can adjust the `flush_bytes`, `flush_interval`, `max_retries` and `timeout` settings described in <<apm-elasticsearch-output,Configure the Elasticsearch output>> to reduce APM Server indexing pressure. However, consider that increasing number of buffered documents and/or reducing retries may lead to a higher rate of dropped APM events. Down below a custom configuration example is listed where the number of default buffered documents is roughly doubled while {es} indexing retries are decreased simultaneously. This configuration provides a generic example and might not be applicable to your situation. Try adjusting the settings further to see what works for you.
```

Author: fixed
